Add custom event handler doc (#1785)
wso2/docs-is: en/docs/develop/writing-a-custom-event-handler.md

3. Publish the `POST_ADD_USER` event — The subscribed handlers will be executed for the post-add user event.

Therefore, the email can be sent through an event handler that is subscribed to the `POST_ADD_USER` event.

The following list is a list of sample events. 

- PRE_AUTHENTICATION
- POST_AUTHENTICATION
- PRE_SET_USER_CLAIMS
- POST_SET_USER_CLAIMS
- PRE_ADD_USER
- POST_ADD_USER
- PRE_UPDATE_CREDENTIAL
- POST_UPDATE_CREDENTIAL
- PRE_UPDATE_CREDENTIAL_BY_ADMIN
- POST_UPDATE_CREDENTIAL_BY_ADMIN
- PRE_DELETE_USER
- POST_DELETE_USER
- PRE_SET_USER_CLAIM
- PRE_GET_USER_CLAIM
- POST_GET_USER_CLAIMS
- POST_GET_USER_CLAIM
- POST_SET_USER_CLAIM
- PRE_DELETE_USER_CLAIMS
- POST_DELETE_USER_CLAIMS
- PRE_DELETE_USER_CLAIM
- POST_DELETE_USER_CLAIM
- PRE_ADD_ROLE
- POST_ADD_ROLE
- PRE_DELETE_ROLE
- POST_DELETE_ROLE
- PRE_UPDATE_ROLE
- POST_UPDATE_ROLE
- PRE_UPDATE_USER_LIST_OF_ROLE
- POST_UPDATE_USER_LIST_OF_ROLE
- PRE_UPDATE_ROLE_LIST_OF_USER
- POST_UPDATE_ROLE_LIST_OF_USER
- UPDATE_GOVERNANCE_CONFIGURATION
- TRIGGER_NOTIFICATION

## Writing an event handler

To write a new event handler, you must extend the `org.wso2.carbon.identity.event.handler.AbstractEventHandler`. 

> 1. Initialize the events in the [Events class](https://github.com/wso2/carbon-identity-framework/blob/0cdd0641a6dd4ece47c79ada5acc05a5d7f9bbe0/components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java#L47) as follows.

Reviewer: This can go in to the publishing custom event and handle that in the handler section.

Reviewer: On second thought, let's remove the initializing events section which is irrelevant to writing an event handler.

USER_STORE_DOMAIN, user.getUserStoreDomain());
  
   Event sampleEvent = new Event(eventName, properties);
   try {
     IdentityRecoveryServiceDataHolder.getInstance(). getIdentityEventService().handleEvent(sampleEvent);
   } catch (IdentityEventException e) {
     throw Utils.handleServerException(ERROR_CODE_TRIGGER_EVENT, e);
}
```

The `handleEvent()` method in the handler class will check for the published event and execute the operation accordingly.

## Registering the event handler 

Register the event handler in the service component as follows.

```
protected void activate(ComponentContext context) {
    try {
        BundleContext bundleContext = context.getBundleContext();
        bundleContext.registerService(AbstractEventHandler.
class.getName(),new SampleEventHandler(), null);
    } catch (Exception e) {
       ...
    }
```

## Configuring the event handler

Add the event handler configuration to the `<IS_HOME>/repository/conf/deployment.toml` file. The events that need to subscribe to the handler can be listed in subscriptions.

```
[[event_handler]]
name= "customEvent"
subscriptions =["PRE_SET_EVENT","POST_SET_EVENT"]
```

> When you want to execute an operation related to an event, publish the event. Then, the handler that is subscribed for the relevant events will be used to execute those events. In the sample configuration given above, the `customEvent` handler is subscribed to `PRE_SET_EVENT` and `POST_SET_EVENT` operations.

Reviewer 2: subscribed to the

# Writing a Custom Event Handler

> The WSO2 Identity Server eventing framework can be used to trigger some events such as user operations. The eventing framework also supports handlers which can be used to do some operations based on the published events.

Reviewer: **....can be used to trigger some events such as user operations.**
Shall we add a couple of sample events, and mention that full list is available below. ex: `can be used to trigger events such as user operation events like PRE_SET_USER_CLAIMS, POST_ADD_USER`

**...framework also supports handlers which can be used to do some operations based on the published events**
Shall we add an example operation as well? ex; `which can be used to do operations upon a triggered event. For example, validate the changed user password against  the previously used entries, upon a PRE_UPDATE_CREDENTIAL event`

- POST_DELETE_USER_CLAIMS
- PRE_DELETE_USER_CLAIM
- POST_DELETE_USER_CLAIM
- PRE_ADD_ROLE
- POST_ADD_ROLE
- PRE_DELETE_ROLE
- POST_DELETE_ROLE
- PRE_UPDATE_ROLE
- POST_UPDATE_ROLE
- PRE_UPDATE_USER_LIST_OF_ROLE
- POST_UPDATE_USER_LIST_OF_ROLE
- PRE_UPDATE_ROLE_LIST_OF_USER
- POST_UPDATE_ROLE_LIST_OF_USER
- UPDATE_GOVERNANCE_CONFIGURATION
- TRIGGER_NOTIFICATION

## Writing an event handler

To write a new event handler, you must extend the `org.wso2.carbon.identity.event.handler.AbstractEventHandler`. 

1. Initialize the events in the [Events class](https://github.com/wso2/carbon-identity-framework/blob/0cdd0641a6dd4ece47c79ada5acc05a5d7f9bbe0/components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java#L47) as follows.

    ```
    public static final String PRE_SET_EVENT = "PRE_SET_EVENT";
    public static final String POST_SET_EVENT = "POST_SET_EVENT";
    ```

1. Override the `getName()` method to set the name for the event handler and the `getPriority()` method can be used to set the priory of the event handler. The handlers will be executed based on the priority.

    ```
    public String getName() {

>         return "customEvent";

Reviewer: ```suggestion
        return "customEventHandler";
```

        return 50;
    }
    ```

3. To execute the expected operation, override the `handleEvent()` method. The `event.getEventProperties()` method can be used to get the parameters related to the user operations. 

    ```
    @Override
    public void handleEvent(Event event) throws IdentityEventException {
    Map<String, Object> eventProperties =event.getEventProperties();
    String userName = (String) eventProperties.get(IdentityEventConstants.EventProperty.USER_NAME);
    UserStoreManager userStoreManager = (UserStoreManager) eventProperties.get(IdentityEventConstants.EventProperty.USER_STORE_MANAGER);
    String tenantDomain = (String) eventProperties.get(IdentityEventConstants.EventProperty.TENANT_DOMAIN);
    String domainName = userStoreManager.getRealmConfiguration().getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_DOMAIN_NAME);
    String eventName = event.getEventName();
    if (IdentityEventConstants.Event.PRE_SET_EVENT.equals(eventName)) {//Do the operation on PRE_SET_EVENT.}
    if(IdentityEventConstants.Event.POST_SET_EVENT.equals(eventName)) {//Do the operation on POST_SET_EVENT.}
    }
    ```

> ## Using the event handler to perform an operation

Reviewer: ```suggestion
## Triggering a custom event and using the event handler to listen and perform an operation
```

Author: Isn't that a bit too long for a subheading? Any way we can make that shorter?

Reviewer: This section needs to go at the end of the document, even after the part where we link available handlers.

Reviewer: On second thought, let's remove the triggering events section which is irrelevant to writing an event handler.

> Write a method to trigger the event. In the method, you can publish an event along with event properties such as `USER_NAME`, `TENANT_DOMAIN`, and `USER_STORE_DOMAIN`, and call the `handleEvent()` method as shown in the following code block.

Reviewer: ```suggestion
Write a method within the product runtime(in an already deployed class file) to trigger an event. In the method, you can publish an event along with event properties such as `USER_NAME`, `TENANT_DOMAIN`, and `USER_STORE_DOMAIN` as shown in the following code block via the `handleEvent()` method in the eventing framework.
```
